mitre/caldera
# [VIRTS-1964] Ability Storage

## Description

Store single abilities, instead of an ability for each platform/executor combination.

## Type of change


## How Has This Been Tested?

Not automatically, yet...

Operations:
  - Manual command
  - Potential link
  - Download operation report

Adversary Profiles:
  - Open page
  - Select adversary: (ex: Worm)
    - Make sure atomic ordering is loaded
    - Make sure key icon shows up (Worm > Copy 54ndc47 (SMB))
    - Make sure lock icon shows up (Worm > Copy 54ndc47 (SMB))
    - Make sure payloads icon shows up (Worm > Copy 54ndc47 (SMB))
    - Make sure cleanup icon shows up (Worm > Copy 54ndc47 (SMB))

  - Open an ability
    - Make sure executors are listed (Worm > Collect ARP details)
    - Make sure payloads are selected (Worm > Copy 54ndc47 (SMB))

  - Add ability
    - Select an ability using dropdowns
    - Search for an ability
    - Create an ability (check everything...)
    - Add ability to adversary

  - Update ability (ex: change command). Make sure parsers are saved.

  - Adversary
    - Save, reload, reorder adversary
    - Load adversary into a new adversary

  - Run an operation with new ability/adversary

Plugins:

  - Access:
    - View link output, including the "Avoid logs" bootstrap ability
    - View ability (should show properties): Discovery > T1033 | System Owner/User Discovery > Current User
    - Search for ability: Current User or "whoami"
    - Modify ability facts: ex: Find user processes
    - Apply ability

  - Builder:
    - Potential link selection shows no errors in console
    - Run a builder-built ability (ex: CredDump.exe, since it requires a DLL as well)

  - Gameboard:
    - External detections:
      - Log in as red
      - Run "Discovery" adversary
      - Open Gameboard
      - Select operation
      - Note PID for "View admin shares" (first ability)
      - Click "Add external detection", enter:
        - tactic: discovery
        - technique: T1135 | Network Share Discovery
        - pid: PID from gameboard
      - See something like: "Detection successfully added for link in operation: test - 2021-03-23 18:32:52.110373."

    - Custom analytic:
      - Create a custom analytic

    - Red vs blue
      - Run red "Discovery" operation
      - Run blue "Incident Responder" operation
      - View points in gameboard

  - Response:
    - Unsure the best way to test

  - Stockpile:
    - Run an ability using donut
      - Add headers to sandcat download: `$wc.Headers.add("gocat-extensions","donut");`
      - Modify `plugins/stockpile/data/abilities/discovery/5a39d7ed-45c9-4a79-b581-e5fb99e24f65.yml`, removing `cmd` and `psh` executors
      - Run "System processes" ability (search "ProcessDump.donut" in potential links)

  - Training:
    - AdversariesFlag2: Follow solution guide for flag

  - Manx:
    - Run a command manually
    - Run a command through an ability

## Checklist:
